File: sample_code/python/simple_streamer/simple_streamer.py
'''
simple_streamer.py

Sample code connect to sensor, capture 20 frames, then display the amplitude image, distance image, and point cloud.
'''

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm
import time
import argparse
import numpy as np
import pytofcore
import open3d as o3d
import cv2
measurement_data=[]
vis = o3d.visualization.Visualizer()
vis.create_window()

def create_cloud_from_distance_image(distance: np.array, rays: np.array, horizontal_binning: bool, vertical_binning: bool):
    # map the distance image (m x n) to a point cloud (m x n x 3)
    slicing = [1, 1]
    if horizontal_binning:
        slicing[0] = 2
    if vertical_binning:
        slicing[1] = 2
    # take subset of rays based on binning configuration
    rays_small = rays[::slicing[1], ::slicing[0], :]
    cloud = distance[..., np.newaxis] * rays_small
    return cloud

def measurement_callback(data):
  if data.data_type == pytofcore.Measurement.DataType.DISTANCE_AMPLITUDE:
    #measurement_data.append(data)
    distance = np.array(data.distance_data)
    amplitude = np.array(data.amplitude_data)
    cv2.imshow("Distance Image",distance*5)
    cv2.imshow("Amplitude Image",amplitude*100)
    cv2.waitKey(1)
    cloud = create_cloud_from_distance_image(distance / 1000.0, rays,False,False)
    x = cloud[:,:,0][amplitude>args.min_amplitude]
    y = cloud[:,:,1][amplitude>args.min_amplitude]
    z = cloud[:,:,2][amplitude>args.min_amplitude]
    amplitude = amplitude[amplitude>args.min_amplitude]
    pcd = o3d.geometry.PointCloud()
    xyz=np.concatenate((z.reshape((-1,1)),x.reshape((-1,1)),y.reshape((-1,1))),axis=1)
    pcd.points = o3d.utility.Vector3dVector(xyz)
    vis.clear_geometries()

    # add pcd to visualizer
    vis.add_geometry(pcd)
    vis.get_view_control().rotate(523.6 * 3, 523.6 * 3, 0, 0)
    vis.get_view_control().set_zoom(0.35)
    vis.get_render_option().point_color_option = o3d.visualization.PointColorOption.YCoordinate

    vis.poll_events()
    vis.update_renderer()
    measurement_callback.counter += 1


parser = argparse.ArgumentParser()
parser.add_argument('--protocol-version', default=pytofcore.Sensor.DEFAULT_PROTOCOL_VERSION, type=int, help="Protocol version to use 0 or 1")
parser.add_argument('-p','--port-name', default=pytofcore.Sensor.DEFAULT_PORT_NAME, type=str, help="Port name to connect to sensor on.")
parser.add_argument('-m','--min_amplitude', default=0, type=int, help="minimum amplitude to display")

args = parser.parse_args()

sensor = pytofcore.Sensor(protocol_version=args.protocol_version, port_name=args.port_name)

rays = np.array(sensor.pixel_rays).reshape(
        (3, 240, 320)).transpose((1, 2, 0))
sensor.subscribe_measurement(measurement_callback)
measurement_callback.counter=0
sensor.set_integration_times(4000,0,0)
sensor.stream_distance_amplitude()
while measurement_callback.counter < 2000:
    time.sleep(0.1)
sensor.stop_stream()

plt.ion()
fig = plt.figure()
ax=[]
ax.append(fig.add_subplot(131))
ax.append(fig.add_subplot(132))
ax.append(fig.add_subplot(133, projection='3d'))
ax[2].view_init(elev=0, azim=180)
ax[0].set_title("Amplitude Image")
ax[1].set_title("Distance Image")
ax[2].set_title("Point Cloud")

#matplotlib is slow for rendering point cloud, try open3d library (used in c++ example) if you want to stream realtime
#for data in measurement_data:
    # distance = np.array(data.distance_data)
    # amplitude = np.array(data.amplitude_data)
    # cv2.imshow("Distance Image",distance)
    # cv2.imshow("Amplitude Image",amplitude)

    # cloud = create_cloud_from_distance_image(distance / 1000.0, rays,False,False)
    # x = cloud[:,:,0][amplitude>args.min_amplitude]
    # y = cloud[:,:,1][amplitude>args.min_amplitude]
    # z = cloud[:,:,2][amplitude>args.min_amplitude]
    # amplitude = amplitude[amplitude>args.min_amplitude]
    # pcd = o3d.geometry.PointCloud()
    # xyz=np.concatenate((x.reshape((-1,1)),y.reshape((-1,1)),z.reshape((-1,1))),axis=1)
    # pcd.points = o3d.utility.Vector3dVector(xyz)
    # vis.clear_geometries()

    # # add pcd to visualizer
    # vis.add_geometry(pcd)
    # vis.poll_events()
    # vis.update_renderer()

File: sample_code/cpp/simple_streamer/simple-streamer.cpp
/**
 * @file simple-streamer.cpp
 *
 * Copyright 2023 PreAct Technologies
 *
 * Test program that uses libtofcore to stream Distance Apmplitude data
 */
#include "tofcore/tof_sensor.hpp"
#include "tofcore/cartesian_transform.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <iostream>
#include <iomanip>
#include <thread>
#include <boost/program_options.hpp>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/opencv.hpp>

#include "open3d/Open3D.h"
#include <Eigen/Dense>

using namespace std::chrono_literals;
using namespace std::chrono;
using namespace tofcore;

static uint32_t baudRate{DEFAULT_BAUD_RATE};
static std::string devicePort{DEFAULT_PORT_NAME};
static volatile bool exitRequested{false};
static uint16_t protocolVersion{DEFAULT_PROTOCOL_VERSION};
static uint16_t minAmplitude{0};

tofcore::CartesianTransform cartesianTransform_;
open3d::visualization::Visualizer vis;


static void measurement_callback(std::shared_ptr<tofcore::Measurement_T> pData)
{

    using DataType = tofcore::Measurement_T::DataType;

    if (pData->type() == DataType::DISTANCE_AMPLITUDE)
    {
        std::vector<Eigen::Vector3d> points = std::vector<Eigen::Vector3d>();
        std::vector<Eigen::Vector3d> look_at = std::vector<Eigen::Vector3d>();

        cv::Mat dist_frame_resized;
        cv::Mat amp_frame_resized;
        // Do filtering on this opencv Mat
        cv::Mat dist_frame = cv::Mat(pData->height(), pData->width(), CV_16UC1, (void *)pData->distance().begin());
        cv::resize(dist_frame, dist_frame_resized, cv::Size(dist_frame.cols * 3, dist_frame.rows * 3));
        dist_frame_resized *= 5; // add some gain to make image look better
        cv::imshow("Distance Image", dist_frame_resized);
        cv::waitKey(1);
        cv::Mat amp_frame = cv::Mat(pData->height(), pData->width(), CV_16UC1, (void *)pData->amplitude().begin());
        cv::resize(amp_frame, amp_frame_resized, cv::Size(amp_frame.cols * 3, amp_frame.rows * 3));
        amp_frame_resized *= 50; // add some gain to make image look better
        cv::imshow("Amplitude Image", amp_frame_resized);
        cv::waitKey(1);
        //cv::medianBlur(dist_frame, dist_frame, 3);
        cv::Mat src = cv::Mat::zeros(dist_frame.size(), CV_32FC1);
        cv::Mat dst = cv::Mat::zeros(dist_frame.size(), CV_32FC1);
        dist_frame.convertTo(src, CV_32FC1);
        cv::bilateralFilter(src, dst, 9, 75,75);
        dst.convertTo(dist_frame, CV_16UC1);
        uint16_t *it_d = (unsigned short *)dist_frame.datastart;
        uint16_t *it_a = (unsigned short *)pData->amplitude().begin();
        int count = 0;
        double px, py, pz;
        uint16_t distance;
        uint16_t y;
        uint16_t x;
        int valid = 0;
        while (it_d != (const unsigned short *)dist_frame.dataend)
        {
            if (*it_a < minAmplitude)
            {
                ++count;
                it_d += 1;
                it_a += 1;
                continue;
            }
            distance = *it_d;
            y = count / pData->width();
            x = count % pData->width();
            valid = 0;
            px = py = pz = 0.1;
            cartesianTransform_.transformPixel(x, y, distance, px, py, pz);
            px /= 1000.0; // mm -> m
            py /= 1000.0; // mm -> m
            pz /= 1000.0; // mm -> m
            auto tmp = Eigen::Vector3d(pz, px, py);
            points.push_back(tmp);
            ++count;
            it_d += 1;
            it_a += 1;
        }
        open3d::geometry::PointCloud point_cloud = open3d::geometry::PointCloud(points);
        std::shared_ptr<open3d::geometry::PointCloud> cloud_ptr = std::make_shared<open3d::geometry::PointCloud>(point_cloud);
        // vis.RemoveGeometry();

        vis.ClearGeometries();
        vis.AddGeometry(cloud_ptr);
        //0.003 [radian/pixel] * (180/pi) [degrees/radian] = 0.1719 [degrees/pixel] OR 5.8178 [pixels/degree]
        //Taking 360 [degrees/rotation] * 5.8178 [pixels/degree] = 2094.3951 [pixels/rotation]
        vis.GetViewControl().Rotate(523.6 * 3, 523.6 * 3, 0, 0);
        vis.GetViewControl().SetZoom(0.25);
        vis.GetRenderOption().point_color_option_ = open3d::visualization::RenderOption::PointColorOption::YCoordinate;
        vis.GetRenderOption().point_size_=3;
        vis.PollEvents();
        vis.UpdateRender();
        vis.UpdateGeometry();
        dist_frame.release();
        dist_frame_resized.release();
        amp_frame.release();
        amp_frame_resized.release();
    }
}
namespace po = boost::program_options;

static void parseArgs(int argc, char *argv[])
{
    po::options_description desc("Simple Streaming Example");
    desc.add_options()("help,h", "produce help message")("device-uri,p", po::value<std::string>(&devicePort)->default_value(devicePort))("protocol-version,v", po::value<uint16_t>(&protocolVersion)->default_value(DEFAULT_PROTOCOL_VERSION))("min-amplitude,m", po::value<uint16_t>(&minAmplitude)->default_value(0));

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);
    if (vm.count("help"))
    {
        std::cout << desc << "\n";
        exit(0);
    }
}

static void signalHandler(int signum)
{
    (void)signum;
    exitRequested = true;
}

int main(int argc, char *argv[])
{
    parseArgs(argc, argv);
    /*
     * Change default action of ^C, ^\ from abnormal termination in order to
     * perform a controlled shutdown.
     */
    signal(SIGINT, signalHandler);
#if defined(SIGQUIT)
    signal(SIGQUIT, signalHandler);
#endif
    {
        tofcore::Sensor sensor{protocolVersion, devicePort, baudRate};
        std::vector<double> rays_x, rays_y, rays_z;
        try
        {
            sensor.getLensInfo(rays_x, rays_y, rays_z);
            cartesianTransform_.initLensTransform(320, 240, rays_x, rays_y, rays_z);
        }
        catch (std::exception &e)
        {
            std::cout << "Error retreiving lens info..." << std::endl;
            return -1;
        }
        vis.CreateVisualizerWindow("Open3D", 1500, 1000, 75, 50, true);
        vis.PollEvents();
        vis.UpdateRender();

        sensor.subscribeMeasurement(&measurement_callback); // callback is called from background thread
        sensor.setIntegrationTime(3000);
        sensor.streamDistanceAmplitude();

        while (!exitRequested) // wait for ^\ or ^C
        {
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        std::cout << "Shutting down..." << std::endl;
        sensor.stopStream();
    } // when scope is exited, sensor connection is cleaned up

    return 0;
}
